### PR TITLE
infra/gcp/aaa: add k8s-infra-prow

### PR DESCRIPTION
**Pull request by ameukam**
Deploy k8s-infra prow instance for configuration focused on k8s.io repo. 

Signed-off-by: Arnaud Meukam <ameukam@gmail.com>

### CODE REVIEW

File: apps/prow/prowjobs/kubernetes/k8s.io/periodics.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+periodics:
```
Comment (spiffxp) [inline apps/prow/prowjobs/kubernetes/k8s.io/periodics.yaml:1]: nit: I am more used to seeing `apps/prow/jobs/{org}/{repo}`

we're gonna want to move this to be top-level or easy to find anyway, so maybe the particular name right now doesn't matter as much

File: apps/prow/plugins.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
+# Plugin configuration
+---
+
+approve:
+- repos:
+  - kubernetes/k8s.io:
```
Comment (spiffxp) [inline apps/prow/plugins.yaml:6]:
```suggestion
  - kubernetes/k8s.io
```

**Review by spiffxp — COMMENTED** [review]
I feel like maybe you grabbed a copy from or were at least inspired by my attempts at setting up github.com/bashfire/prow-config.  The problem is that's fallen way stale at this point.  I would recommend a refresh from https://github.com/kubernetes/test-infra/tree/master/config/prow just to be sure

I dropped some notes to self that I'll try and answer as followup

File: apps/prow/cluster/prow_ingress.yaml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+metadata:
+  name: k8s-infra-prow-k8s-io
+  namespace: prow
+  annotations:
+    kubernetes.io/ingress.class: "gce"
+    kubernetes.io/ingress.global-static-ip-name: k8s-infra-prow
+    networking.gke.io/managed-certificates: k8s-infra-prow-k8s-io
+spec:
+  rules:
+  - host: k8s-infra-prow.k8s.io
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/cluster/prow_ingress.yaml:12]: Created in #2404.

File: apps/prow/cluster/deck_deployment.yaml [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        secret:
+          defaultMode: 420
+          secretName: github-oauth-config
+      - name: github-token
+        secret:
+          defaultMode: 420
+          secretName: k8s-infra-ci-robot-github-token
+      - name: cookie-secret
+        secret:
+          secretName: cookie
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/cluster/deck_deployment.yaml]: Added secret to GCP Secrets Manager and ExternalSecret Matching it.

File: apps/prow/cluster/prow-externalsecrets.yaml [hunk 5]
```diff
@@ -32,3 +32,19 @@ spec:
   - key: k8s-infra-build-clusters-kubeconfig # The name of the GSM Secret
     name: kubeconfig                         # The key to write to in the Kubernetes Secret
     version: latest                          # The version of the GSM Secret
+---
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/cluster/prow-externalsecrets.yaml:51]: ExternalSecret added.

File: apps/prow/cluster/tide_deployment.yaml [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      - name: tide
+        image: gcr.io/k8s-prow/tide:v20210407-51f95c2d52
+        args:
+        - --config-path=/etc/config/config.yaml
+        - --dry-run=false
+        - --github-endpoint=http://ghproxy.prow.svc.cluster.local
+        - --github-endpoint=https://api.github.com
+        - --github-token-path=/etc/github/token
+        - --job-config-path=/etc/job-config
+        - --history-uri=gs://k8s-infra-prow-logs/tide-history.json
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/cluster/tide_deployment.yaml]: Switched to `k8s-infra-prow-results`.

File: apps/prow/config.yaml [hunk 7]
```diff
@@ -0,0 +1,115 @@
+# config.yaml
+prowjob_namespace: prow
+pod_namespace: k8s-infra-test-pods
+log_level: debug
+
+plank:
+  job_url_template: 'https://k8s-infra-prow.k8s.io/view/gcs/bashfire-prow/{{if eq .Spec.Type "presubmit"}}pr-logs/pull{{else if eq .Spec.Type "batch"}}pr-logs/pull{{else}}logs{{end}}{{if .Spec.Refs}}{{if ne .Spec.Refs.Org ""}}/{{.Spec.Refs.Org}}_{{.Spec.Refs.Repo}}{{end}}{{end}}{{if eq .Spec.Type "presubmit"}}/{{with index .Spec.Refs.Pulls 0}}{{.Number}}{{end}}{{else if eq .Spec.Type "batch"}}/batch{{end}}/{{.Spec.Job}}/{{.Status.BuildID}}/'
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/config.yaml]: Fixed.

File: apps/prow/config.yaml [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+  report_templates:
+    '*': '[Full PR test history](https://k8s-infra-prow.k8s.io/pr-history?org={{.Spec.Refs.Org}}&repo={{.Spec.Refs.Repo}}&pr={{with index .Spec.Refs.Pulls 0}}{{.Number}}{{end}}). [Your PR dashboard](https://k8s-infra-prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3A{{with index .Spec.Refs.Pulls 0}}{{.Author}}{{end}}).'
+  job_url_prefix_config:
+    '*': https://k8s-infra-prow.k8s.io/view/gcs/
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/config.yaml]: Fixed.

File: apps/prow/config.yaml [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+tide:
+  sync_period: 1m
+  queries:
+  - orgs:
+    - kubernetes
+    labels:
+    - lgtm
+    missingLabels:
+    - do-not-merge/hold
+    - do-not-merge/work-in-progress
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/config.yaml]: Fixed.

File: apps/prow/config.yaml [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+  pr_status_base_urls:
+    '*': https://k8s-infra-prow.k8s.io/pr
+
+deck:
+  branding:
+    logo: 'https://raw.githubusercontent.com/cncf/artwork/master/projects/kubernetes/icon/color/kubernetes-icon-color.svg'
+    favicon: 'https://kubernetes.io/images/favicon.png'
+  spyglass:
+    size_limit: 100000000 # 100MB
+    gcs_browser_prefix: https://gcsweb.k8s.io/gcs/
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/config.yaml:76]: Applied

File: apps/prow/plugins.yaml [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    - lgtm
+    - lifecycle
+    # - owners-label
+    - shrug
+    - size
+    - skip
+    - trigger
+    - wip
+    # - verify-owners
+    - yuks
```
[the remark this answers is not in the document]
Comment (ameukam) [reply apps/prow/plugins.yaml:55]: The suggestion looks the same. (It's kind of late for me).

File: apps/prow/prowjobs/kubernetes/k8s.io/periodics.yaml (hunk 1 above)
Comment (ameukam) [reply apps/prow/prowjobs/kubernetes/k8s.io/periodics.yaml:1]: We'll update location when I'll add prow self-maintenance jobs.

File: apps/prow/plugins.yaml (hunk 2 above)
Comment (ameukam) [reply apps/prow/plugins.yaml]: Applied.

File: apps/prow/plugins.yaml (hunk 11 above)
[the remark this answers is not in the document]
Comment (spiffxp) [reply apps/prow/plugins.yaml:55]: I dropped the suggestion, you're right it was the same and I think I talked myself into "well, let's try it out and see what happens"

**Review by spiffxp — CHANGES_REQUESTED** [review]
one nit, just the one, otherwise it LGTM

**Comment by ameukam** [comment]
> one nit, just the one, otherwise it LGTM

@spiffxp removed the lines.

**Review by spiffxp — APPROVED** [review]
/approve
/lgtm

**Comment by spiffxp** [comment]
This failed to auto-deploy because RBAC:
- Open issue is: https://github.com/kubernetes/k8s.io/issues/2218
- PR to workaround is: https://github.com/kubernetes/k8s.io/pull/2483

**Comment by spiffxp** [comment]
Retriggered job via re-run button https://prow.k8s.io/?job=post-k8sio-deploy-app-prow

**Comment by spiffxp** [comment]
OK I think the remaining errors are token / typo related https://prow.k8s.io/view/gs/kubernetes-jenkins/logs/post-k8sio-deploy-app-prow/1423707644174012416

will leave to @ameukam to fix for now

**Comment by ameukam** [comment]
Some issues we have post-deployment : 

- deck need to allowed to push objects on `gs://k8s-testgrid`: 

```shell
kubectl -n prow logs -l app=deck
...
{"component":"deck","error":"Get \"https://storage.googleapis.com/k8s-testgrid/config\": compute: Received 403 `Unable to generate access token; IAM returned 403 Forbidden: The caller does not have permission\nThis error could be caused by a missing IAM policy binding on the target IAM service account.\nFor more information, refer to the Workload Identity documentation:\n\thttps://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to\n\n`","file":"prow/spyglass/testgrid.go:50","func":"k8s.io/test-infra/prow/spyglass.(*TestGrid).Start.func1","level":"error","msg":"Couldn't update TestGrid config.","path":"gs://k8s-testgrid/config","severity":"error","time":"2021-08-06T21:50:08Z"}
```
@spiffxp I believe this is under a Google GCP Org. IAM binding may need to be done by Googlers.

- tide need to push objects to `gs://k8s-infra-prow-results`. 

```shell
{"component":"tide","error":"error initializing history client from \"gs://k8s-infra-prow-results/tide-history.json\": open: Get \"https://storage.googleapis.com/k8s-infra-prow-results/tide-history.json\": compute: Received 403 `Unable to generate access token; IAM returned 403 Forbidden: The caller does not have permission\nThis error could be caused by a missing IAM policy binding on the target IAM service account.\nFor more information, refer to the Workload Identity documentation:\n\thttps://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#authenticating_to\n\n`","file":"prow/cmd/tide/main.go:176","func":"main.main","level":"fatal","msg":"Error creating Tide controller.","severity":"fatal","time":"2021-08-06T22:55:38Z"}
```

**Comment by ameukam** [comment]
Ref: https://github.com/kubernetes/k8s.io/issues/1394

**Pull request merged**